File: cardgrader/export.py
"""Render the annotated overlay image.

The export deliberately matches what Home shows on screen: the card
straightened by the stage-1 angle, cropped to the outer frame with the same
margin around it, rather than the whole scan with its bed and shadows.  A
caption band underneath carries the numbers, so the picture is self-contained
when it is sent to somebody else.
"""

from __future__ import annotations

import os

import numpy as np
from PIL import Image, ImageDraw, ImageFont

from . import grades
from .geometry import Line, clip_line_to_rect, vec
from .model import INNER, OUTER
from .palette import SAMPLE_C, color_for, scheme_colors
from .tint import apply_tint
from .transform import ViewTransform

BG = (30, 31, 34)
BAND = (22, 23, 26)
TEXT = (230, 232, 234)
MUTED = (154, 160, 166)
RULE = (58, 62, 68)

# Same fraction of the frame the card fills under Home.
HOME_FILL = 0.88

_FONT_CANDIDATES = (
    "segoeui.ttf", "arial.ttf", "calibri.ttf",          # Windows
    "/System/Library/Fonts/Supplemental/Arial.ttf",     # macOS
    "DejaVuSans.ttf",                                   # Linux / Pillow
)
_FONT_BOLD_CANDIDATES = (
    "segoeuib.ttf", "arialbd.ttf", "calibrib.ttf",
    "/System/Library/Fonts/Supplemental/Arial Bold.ttf",
    "DejaVuSans-Bold.ttf",
)


def _font(size: int, bold: bool = False):
    for name in (_FONT_BOLD_CANDIDATES if bold else _FONT_CANDIDATES):
        try:
            return ImageFont.truetype(name, size)
        except Exception:
            continue
    try:
        return ImageFont.load_default(size)
    except TypeError:                      # Pillow < 10 has no size argument
        return ImageFont.load_default()


def _text_width(draw, text, font):
    try:
        box = draw.textbbox((0, 0), text, font=font)
        return box[2] - box[0]
    except Exception:
        return draw.textlength(text, font=font)


def build_transform(model, fill=HOME_FILL):
    """Export transform plus output size: card upright, Home's margin around it.

    Returns (transform, (width, height)) or None if the frame is degenerate.
    """
    corners = model.outer_corners()
    if corners is None:
        return None
    t = ViewTransform(angle=model.transform.angle, scale=1.0,
                      pan=vec(0.0, 0.0),
                      center=model.transform.center.copy())
    view = np.array([t.img_to_view(p) for p in corners], dtype=float)
    x0, y0 = view[:, 0].min(), view[:, 1].min()
    x1, y1 = view[:, 0].max(), view[:, 1].max()
    cw, ch = x1 - x0, y1 - y0
    if cw < 2 or ch < 2:
        return None
    margin_x = cw * (1.0 / fill - 1.0) / 2.0
    margin_y = ch * (1.0 / fill - 1.0) / 2.0
    t.pan = vec(margin_x - x0, margin_y - y0)
    return t, (int(round(cw + 2 * margin_x)), int(round(ch + 2 * margin_y)))


def render_card(model, scheme=None, tint_mode="None", tint_strength=0.0,
                show_samples=True):
    """The card alone: straightened, cropped to Home's framing, lines drawn."""
    built = build_transform(model)
    if built is None:                       # no usable frame: fall back to the scan
        t = ViewTransform(angle=model.transform.angle, scale=1.0,
                          pan=vec(model.size[0] / 2.0, model.size[1] / 2.0),
                          center=model.transform.center.copy())
        size = model.size
    else:
        t, size = built

    card = model.image.transform(size, Image.AFFINE, t.affine_coeffs(),
                                 resample=Image.BICUBIC, fillcolor=BG)
    card = apply_tint(card, tint_mode, tint_strength)
    d = ImageDraw.Draw(card)

    m = model.measure()
    if show_samples and m and m.valid:
        for axis in (m.lr, m.tb):
            for (pa, a, b, pb) in axis.sample_points:
                va, vb = t.img_to_view(pa), t.img_to_view(pb)
                d.line([tuple(va), tuple(vb)], fill=SAMPLE_C, width=1)
                for q in (pa, a, b, pb):
                    vq = t.img_to_view(q)
                    d.ellipse([vq[0] - 3, vq[1] - 3, vq[0] + 3, vq[1] + 3],
                              fill=SAMPLE_C)

    width = max(2, int(round(size[0] / 500.0)))
    for key in list(OUTER) + list(INNER):
        line = model.lines[key]
        vp = Line(t.img_to_view(line.p1), t.img_to_view(line.p2))
        span = clip_line_to_rect(vp, 0, 0, size[0] - 1, size[1] - 1)
        if span:
            a, b = span
            d.line([tuple(a), tuple(b)],
                   fill=color_for(key, scheme), width=width)

    return card


def render_overlay(model, scheme=None, tint_mode="None", tint_strength=0.0,
                   side=grades.FRONT, show_samples=True, caption=True):
    """One side, with the caption band under it."""
    card = render_card(model, scheme, tint_mode, tint_strength, show_samples)
    if not caption:
        return card
    return _add_caption(card, model, model.measure(), scheme, side)


def _add_caption(card, model, m, scheme, side):
    w = card.width
    unit = max(11, int(round(w / 46.0)))            # scales with the card
    pad = unit
    band_h = int(unit * 9.4)
    out = Image.new("RGB", (w, card.height + band_h), BAND)
    out.paste(card, (0, 0))
    d = ImageDraw.Draw(out)
    d.line([(0, card.height), (w, card.height)], fill=RULE, width=1)

    f_small = _font(int(unit * 0.85))
    f_label = _font(int(unit * 0.8))
    f_big = _font(int(unit * 2.0), bold=True)
    f_grade = _font(int(unit * 1.15), bold=True)

    y = card.height + pad
    name = os.path.basename(model.path) if model.path else "card"
    d.text((pad, y), name, font=f_small, fill=MUTED)
    n = m.n_samples if m else model.n_samples
    right = "%d samples/axis   %s" % (n, side)
    d.text((w - pad - _text_width(d, right, f_small), y), right,
           font=f_small, fill=MUTED)
    y += int(unit * 1.5)

    if m is None or not m.valid:
        d.text((pad, y), "no valid measurement", font=f_big, fill=TEXT)
        return out

    colors = scheme_colors(scheme)
    col2 = w // 2
    for x, label, value, colour in (
            (pad, "LEFT / RIGHT", m.lr.text(), colors["outer"]),
            (col2, "TOP / BOTTOM", m.tb.text(), colors["inner"])):
        d.text((x, y), label, font=f_label, fill=MUTED)
        d.text((x, y + int(unit * 1.05)), value, font=f_big, fill=TEXT)
        bar_y = y + int(unit * 3.35)
        bar_w = col2 - pad * 2
        pct = (m.lr if x == pad else m.tb).low_pct
        split = int(bar_w * pct / 100.0)
        d.rectangle([x, bar_y, x + split, bar_y + int(unit * 0.5)], fill=colour)
        d.rectangle([x + split, bar_y, x + bar_w, bar_y + int(unit * 0.5)],
                    fill=RULE)
        d.line([(x + bar_w // 2, bar_y), (x + bar_w // 2, bar_y + int(unit * 0.5))],
               fill=(255, 255, 255))

    y += int(unit * 4.3)
    if m.crossed:
        d.text((pad, y), "outer line sits inside its inner line -- not measurable",
               font=f_grade, fill=(255, 82, 82))
        return out

    x = pad
    for c in grades.ceilings(m, side):
        d.text((x, y), c.grader, font=f_small, fill=MUTED)
        d.text((x + _text_width(d, c.grader + " ", f_small), y - int(unit * 0.12),),
               c.grade, font=f_grade, fill=TEXT)
        x += int(unit * 5.0)
    note = "centering ceiling only -- corners, edges and surface not assessed"
    d.text((w - pad - _text_width(d, note, f_small), y + int(unit * 0.2)), note,
           font=f_small, fill=MUTED)

    spread = max(m.lr.spread, m.tb.spread)
    if spread > 4.0:
        d.text((pad, y + int(unit * 1.6)),
               "samples disagree by %.1f%% -- frames are not parallel" % spread,
               font=f_small, fill=colors["inner"])
    return out


# ------------------------------------------------------------ both sides
def render_session(session, scheme=None, tint_mode="None", tint_strength=0.0,
                   show_samples=True):
    """A single sheet for the whole session.

    With both faces measured they are placed side by side under one caption
    carrying each side's numbers and the final ceiling; with only one face
    loaded this is just that side's overlay.
    """
    loaded = session.loaded_sides()
    if not loaded:
        return None
    if len(loaded) == 1:
        side = loaded[0]
        return render_overlay(session.cards[side], scheme, tint_mode,
                              tint_strength, side, show_samples)

    cards = {}
    for side in loaded:
        cards[side] = render_card(session.cards[side], scheme, tint_mode,
                                  tint_strength, show_samples)
    # Match heights by shrinking the taller one, never by upscaling.
    height = min(c.height for c in cards.values())
    for side, img in cards.items():
        if img.height != height:
            w = max(1, int(round(img.width * height / img.height)))
            cards[side] = img.resize((w, height), Image.LANCZOS)

    gap = max(10, int(height * 0.015))
    margin = gap
    total_w = sum(c.width for c in cards.values()) + gap + margin * 2
    row = Image.new("RGB", (total_w, height + margin), BAND)
    x = margin
    columns = {}
    for side in loaded:
        row.paste(cards[side], (x, margin // 2))
        columns[side] = (x, cards[side].width)
        x += cards[side].width + gap
    return _add_pair_caption(row, session, scheme, columns, margin)


def _add_pair_caption(row, session, scheme, columns, margin):
    w = row.width
    unit = max(11, int(round(w / 78.0)))
    pad = unit

    # Stack the caption explicitly rather than by fractions of the band, so
    # nothing collides when the sheet is small.
    y_label = 0
    y_lr = int(unit * 1.5)
    y_tb = int(unit * 3.4)
    y_note = int(unit * 5.4)
    y_rule = int(unit * 7.1)
    y_final = int(unit * 7.9)
    band_h = pad + int(unit * 10.4)

    out = Image.new("RGB", (w, row.height + band_h), BAND)
    out.paste(row, (0, 0))
    d = ImageDraw.Draw(out)
    d.line([(0, row.height), (w, row.height)], fill=RULE, width=1)

    f_small = _font(int(unit * 0.85))
    f_side = _font(int(unit * 0.95), bold=True)
    f_big = _font(int(unit * 1.7), bold=True)
    f_final = _font(int(unit * 2.1), bold=True)

    colors = scheme_colors(scheme)
    top = row.height + pad

    for side, (x, cw) in columns.items():
        m = session.measurement(side)
        d.text((x, top + y_label), side.upper(), font=f_side, fill=TEXT)
        if m is None or not m.valid:
            d.text((x, top + y_lr), "--", font=f_big, fill=MUTED)
            continue
        d.text((x, top + y_lr), "L/R  " + m.lr.text(),
               font=f_big, fill=colors["outer"])
        d.text((x, top + y_tb), "T/B  " + m.tb.text(),
               font=f_big, fill=colors["inner"])
        spread = max(m.lr.spread, m.tb.spread)
        note = "%d samples/axis" % m.n_samples
        if spread > 4.0:
            note += "    samples disagree by %.0f%%" % spread
        d.text((x, top + y_note), note, font=f_small, fill=MUTED)

    d.line([(margin, top + y_rule), (w - margin, top + y_rule)],
           fill=RULE, width=1)
    d.text((margin, top + y_final + int(unit * 0.55)), "FINAL CEILING",
           font=f_small, fill=MUTED)

    x = margin + int(unit * 8.5)
    for c in session.final_ceilings():
        d.text((x, top + y_final + int(unit * 0.55)), c.grader,
               font=f_small, fill=MUTED)
        d.text((x + int(unit * 2.4), top + y_final), c.grade,
               font=f_final, fill=TEXT)
        x += int(unit * 6.6)

    notes = ["centering only -- corners, edges and surface not assessed"]
    limiting = session.limiting_side()
    if limiting:
        notes.insert(0, "held down by the %s" % limiting)
    if not session.both_loaded:
        notes.insert(0, "only the %s was measured"
                     % ", ".join(session.loaded_sides()))
    for i, text in enumerate(notes):
        d.text((w - margin - _text_width(d, text, f_small),
                top + y_final + i * int(unit * 1.2)),
               text, font=f_small, fill=MUTED)
    return out


def save_session_sheet(session, path, **kw) -> None:
    img = render_session(session, **kw)
    if img is not None:
        img.save(path)


def save_overlay(model, path, **kw) -> None:
    render_overlay(model, **kw).save(path)

File: src/ui/export.ts
/**
 * Annotated overlay images.
 *
 * Each card is drawn the way Home frames it -- upright by its stage-1 angle,
 * cropped to the outer frame with the same margin -- rather than as the whole
 * scan. A caption band carries the numbers so the picture stands on its own
 * when it is sent to somebody. With both faces measured they share one sheet
 * under a caption with the final ceiling.
 */

import { exportFrame } from "../core/framing";
import { type Vec, Line, clipLineToRect } from "../core/geometry";
import { type Side, ceilings } from "../core/grades";
import type { Measurement } from "../core/measure";
import { type CardModel, INNER, OUTER } from "../core/model";
import { type TintMode, applyTint } from "../core/raster";
import type { Session } from "../core/session";
import { ViewTransform } from "../core/transform";
import { rasterToCanvas } from "../platform/files";
import { SAMPLE_C, type SchemeName, colorFor, schemeColors } from "./palette";

const BG = "#1e1f22";
const BAND = "#16171a";
const TEXT = "#e6e8ea";
const MUTED = "#9aa0a6";
const RULE = "#3a3e44";
const FONT = "system-ui, -apple-system, 'Segoe UI', Roboto, sans-serif";

export interface ExportLook {
  scheme: SchemeName;
  tintMode: TintMode;
  tintStrength: number;
  showSamples: boolean;
}

function newCanvas(w: number, h: number): [HTMLCanvasElement, CanvasRenderingContext2D] {
  const c = document.createElement("canvas");
  c.width = Math.max(1, Math.round(w));
  c.height = Math.max(1, Math.round(h));
  return [c, c.getContext("2d")!];
}

/** The card alone: straightened, cropped to Home's framing, lines drawn. */
export function renderCard(model: CardModel, look: ExportLook): HTMLCanvasElement {
  const img = model.image!;
  const built = exportFrame(model);
  const t = built?.transform
    ?? new ViewTransform(model.transform.angle, 1, [img.width / 2, img.height / 2], [...model.transform.center]);
  const size: Vec = built?.size ?? [img.width, img.height];

  const [canvas, ctx] = newCanvas(size[0], size[1]);
  ctx.fillStyle = BG;
  ctx.fillRect(0, 0, canvas.width, canvas.height);
  const src = rasterToCanvas(applyTint(img, look.tintMode, look.tintStrength));
  ctx.save();
  ctx.setTransform(...t.canvasMatrix());
  ctx.imageSmoothingQuality = "high";
  ctx.drawImage(src, 0, 0);
  ctx.restore();

  const m = model.measure();
  if (look.showSamples && m?.valid) {
    ctx.strokeStyle = SAMPLE_C;
    ctx.fillStyle = SAMPLE_C;
    ctx.lineWidth = 1;
    for (const axis of [m.lr, m.tb]) {
      for (const pts of axis.samplePoints) {
        const va = t.imgToView(pts[0]);
        const vb = t.imgToView(pts[3]);
        ctx.beginPath(); ctx.moveTo(va[0], va[1]); ctx.lineTo(vb[0], vb[1]); ctx.stroke();
        for (const q of pts) {
          const v = t.imgToView(q);
          ctx.beginPath(); ctx.arc(v[0], v[1], 3, 0, Math.PI * 2); ctx.fill();
        }
      }
    }
  }

  ctx.lineWidth = Math.max(2, Math.round(size[0] / 500));
  for (const key of [...OUTER, ...INNER]) {
    const l = model.lines![key];
    const seg = clipLineToRect(new Line(t.imgToView(l.p1), t.imgToView(l.p2)), 0, 0, size[0] - 1, size[1] - 1);
    if (!seg) continue;
    ctx.strokeStyle = colorFor(key, look.scheme);
    ctx.beginPath(); ctx.moveTo(seg[0][0], seg[0][1]); ctx.lineTo(seg[1][0], seg[1][1]); ctx.stroke();
  }
  return canvas;
}

const font = (px: number, weight = 400) => `${weight} ${Math.round(px)}px ${FONT}`;

function textRight(ctx: CanvasRenderingContext2D, s: string, right: number, y: number): void {
  ctx.fillText(s, right - ctx.measureText(s).width, y);
}

function bar(ctx: CanvasRenderingContext2D, x: number, y: number, w: number, h: number, pct: number, colour: string): void {
  const split = Math.round((w * pct) / 100);
  ctx.fillStyle = colour;
  ctx.fillRect(x, y, split, h);
  ctx.fillStyle = RULE;
  ctx.fillRect(x + split, y, w - split, h);
  ctx.fillStyle = "#ffffff";
  ctx.fillRect(x + Math.round(w / 2), y, 1, h);
}

/** One face with the caption band underneath. */
export function renderOverlay(model: CardModel, side: Side, look: ExportLook): HTMLCanvasElement {
  const card = renderCard(model, look);
  const m = model.measure();
  const w = card.width;
  const unit = Math.max(11, Math.round(w / 46));
  const pad = unit;
  const bandH = Math.round(unit * 9.6);
  const [out, ctx] = newCanvas(w, card.height + bandH);
  ctx.fillStyle = BAND;
  ctx.fillRect(0, 0, out.width, out.height);
  ctx.drawImage(card, 0, 0);
  ctx.fillStyle = RULE;
  ctx.fillRect(0, card.height, w, 1);
  ctx.textBaseline = "top";

  let y = card.height + pad;
  ctx.fillStyle = MUTED;
  ctx.font = font(unit * 0.85);
  ctx.fillText(model.name ?? "card", pad, y);
  textRight(ctx, `${model.nSamples} samples/axis   ${side}`, w - pad, y);
  y += unit * 1.5;

  if (!m?.valid) {
    ctx.fillStyle = TEXT;
    ctx.font = font(unit * 2, 700);
    ctx.fillText("no valid measurement", pad, y);
    return out;
  }
  const cols = schemeColors(look.scheme);
  const col2 = Math.round(w / 2);
  const axes: Array<[number, string, typeof m.lr, string]> = [
    [pad, "LEFT / RIGHT", m.lr, cols.outer], [col2, "TOP / BOTTOM", m.tb, cols.inner],
  ];
  for (const [x, label, axis, colour] of axes) {
    ctx.fillStyle = MUTED;
    ctx.font = font(unit * 0.8);
    ctx.fillText(label, x, y);
    ctx.fillStyle = TEXT;
    ctx.font = font(unit * 2, 700);
    ctx.fillText(axis.text(), x, y + unit * 1.05);
    bar(ctx, x, Math.round(y + unit * 3.5), col2 - pad * 2, Math.round(unit * 0.5), axis.lowPct, colour);
  }
  y += unit * 4.5;

  if (m.crossed) {
    ctx.fillStyle = "#ff5252";
    ctx.font = font(unit * 1.15, 700);
    ctx.fillText("an outer line sits inside its inner line — not measurable", pad, y);
    return out;
  }
  let x = pad;
  for (const c of ceilings(m, side)) {
    ctx.fillStyle = MUTED;
    ctx.font = font(unit * 0.85);
    ctx.fillText(c.grader, x, y + unit * 0.25);
    ctx.fillStyle = TEXT;
    ctx.font = font(unit * 1.2, 700);
    ctx.fillText(c.grade, x + unit * 2.4, y);
    x += unit * 5.2;
  }
  ctx.fillStyle = MUTED;
  ctx.font = font(unit * 0.8);
  textRight(ctx, "centering ceiling only — corners, edges and surface not assessed", w - pad, y + unit * 0.3);
  const spread = Math.max(m.lr.spread, m.tb.spread);
  if (spread > 4) {
    ctx.fillStyle = cols.inner;
    ctx.fillText(`samples disagree by ${spread.toFixed(1)}% — frames are not parallel`, pad, y + unit * 1.8);
  }
  return out;
}

function sideSummary(ctx: CanvasRenderingContext2D, m: Measurement | null, side: Side, n: number,
  x: number, top: number, unit: number, look: ExportLook): void {
  const cols = schemeColors(look.scheme);
  ctx.fillStyle = TEXT;
  ctx.font = font(unit * 0.95, 700);
  ctx.fillText(side.toUpperCase(), x, top);
  if (!m?.valid) {
    ctx.fillStyle = MUTED;
    ctx.font = font(unit * 1.7, 700);
    ctx.fillText("--", x, top + unit * 1.5);
    return;
  }
  ctx.font = font(unit * 1.7, 700);
  ctx.fillStyle = cols.outer;
  ctx.fillText(`L/R  ${m.lr.text()}`, x, top + unit * 1.5);
  ctx.fillStyle = cols.inner;
  ctx.fillText(`T/B  ${m.tb.text()}`, x, top + unit * 3.4);
  const spread = Math.max(m.lr.spread, m.tb.spread);
  ctx.fillStyle = MUTED;
  ctx.font = font(unit * 0.85);
  ctx.fillText(`${n} samples/axis${spread > 4 ? `    samples disagree by ${spread.toFixed(0)}%` : ""}`, x, top + unit * 5.4);
}

/** One sheet for the whole session: both faces side by side, or one alone. */
export function renderSession(session: Session, look: ExportLook): HTMLCanvasElement | null {
  const loaded = session.loadedSides();
  if (!loaded.length) return null;
  if (loaded.length === 1) return renderOverlay(session.cards[loaded[0]], loaded[0], look);

  const cards = loaded.map((s) => renderCard(session.cards[s], look));
  const height = Math.min(...cards.map((c) => c.height)); // shrink, never upscale
  const widths = cards.map((c) => Math.round((c.width * height) / c.height));
  const gap = Math.max(10, Math.round(height * 0.015));
  const margin = gap;
  const w = widths[0] + widths[1] + gap + margin * 2;
  const unit = Math.max(11, Math.round(w / 78));
  const pad = unit;
  const rowH = height + margin;
  const bandH = pad + Math.round(unit * 10.6);

  const [out, ctx] = newCanvas(w, rowH + bandH);
  ctx.fillStyle = BAND;
  ctx.fillRect(0, 0, out.width, out.height);
  ctx.imageSmoothingQuality = "high";
  const xs = [margin, margin + widths[0] + gap];
  cards.forEach((c, i) => ctx.drawImage(c, xs[i], Math.round(margin / 2), widths[i], height));
  ctx.fillStyle = RULE;
  ctx.fillRect(0, rowH, w, 1);
  ctx.textBaseline = "top";

  const top = rowH + pad;
  loaded.forEach((s, i) => sideSummary(ctx, session.measurement(s), s, session.cards[s].nSamples, xs[i], top, unit, look));

  const yRule = top + Math.round(unit * 7.1);
  const yFinal = top + Math.round(unit * 7.9);
  ctx.fillStyle = RULE;
  ctx.fillRect(margin, yRule, w - margin * 2, 1);
  ctx.fillStyle = MUTED;
  ctx.font = font(unit * 0.85);
  ctx.fillText("FINAL CEILING", margin, yFinal + unit * 0.55);
  let x = margin + unit * 8.5;
  for (const c of session.finalCeilings()) {
    ctx.fillStyle = MUTED;
    ctx.font = font(unit * 0.85);
    ctx.fillText(c.grader, x, yFinal + unit * 0.55);
    ctx.fillStyle = TEXT;
    ctx.font = font(unit * 2.1, 700);
    ctx.fillText(c.grade, x + unit * 2.4, yFinal);
    x += unit * 6.6;
  }
  const notes = ["centering only — corners, edges and surface not assessed"];
  const limiting = session.limitingSide();
  if (limiting) notes.unshift(`held down by the ${limiting}`);
  ctx.fillStyle = MUTED;
  ctx.font = font(unit * 0.85);
  notes.forEach((s, i) => textRight(ctx, s, w - margin, yFinal + i * unit * 1.2));
  return out;
}
